src/middlewared/middlewared/plugins/dlm.py

```diff
@@ -0,0 +1,298 @@
+                return
+
+            # Make config changes
+            await self.middleware.call('dlm.kernel.lockspace_remove_node', lockspace_name, leaving_nodeid)
+
+        elif await self.middleware.call('failover.remote_connected'):
+            # Remote operation
+            self.logger.debug('[REMOTE] Node %s leaving kernel lockspace %s on %s', leaving_nodeid, lockspace_name, dest_nodeid)
+            await self.middleware.call(
+                'failover.call_remote', 'dlm.leave_kernel_lockspace', [dest_nodeid, lockspace_name, leaving_nodeid], {'timeout': 5}
+            )
+
+    @private
+    async def join_lockspace(self, lockspace_name):
+        self.logger.info('Joining lockspace %s', lockspace_name)
+        await self.middleware.call('dlm.create')
+        try:
+            # Note that by virtue of this being a join_lockspace kernel lockspace stopped is already True (on this node)
+            await self.middleware.call('dlm.kernel.lockspace_mark_stopped', lockspace_name)
+
+            nodeIDs = set(await self.middleware.call('dlm.lockspace_members', lockspace_name))
+
+            # Stop kernel lockspace (on all other nodes)
+            await asyncio.gather(*[self.stop_kernel_lockspace(nodeid, lockspace_name) for nodeid in nodeIDs])
+
+            nodeIDs.add(self.nodeID)
+            # Join the kernel lockspace (on all nodes)
+            await asyncio.gather(*[self.join_kernel_lockspace(nodeid, lockspace_name, self.nodeID, list(nodeIDs)) for nodeid in nodeIDs])
+        except Exception:
+            self.logger.error('Failed to join lockspace %s', lockspace_name, exc_info=True)
+            await self.middleware.call('dlm.kernel.set_sysfs_event_done', lockspace_name, 1)
+
+    @private
+    async def leave_lockspace(self, lockspace_name):
+        self.logger.info('Leaving lockspace %s', lockspace_name)
+        await self.middleware.call('dlm.create')
+        try:
+
+            nodeIDs = set(await self.middleware.call('dlm.lockspace_members', lockspace_name))
+
+            # Stop kernel lockspace (on all nodes)
+            await asyncio.gather(*[self.stop_kernel_lockspace(nodeid, lockspace_name) for nodeid in nodeIDs])
+
+            # Leave the kernel lockspace (on all nodes).
+            await asyncio.gather(*[self.leave_kernel_lockspace(nodeid, lockspace_name, self.nodeID) for nodeid in nodeIDs])
+
+            nodeIDs.remove(self.nodeID)
+            # Start the kernel lockspace on remaining nodes
+            await asyncio.gather(*[self.start_kernel_lockspace(nodeid, lockspace_name) for nodeid in nodeIDs])
+
+        except Exception:
+            self.logger.error('Failed to leave lockspace %s', lockspace_name, exc_info=True)
+            await self.middleware.call('dlm.kernel.lockspace_start', lockspace_name)
+            await self.middleware.call('dlm.kernel.set_sysfs_event_done', lockspace_name, 1)
+
+    @private
+    async def add_node(self, nodeid):
```

Comment:
I think it's okay to leave place-holder methods for future enhancements, but we probably need to note that it's a placeholder and raise a NonImplementedError at beginning of method